[CORL-3213] premoderate and ban aliased email addresses under certain conditions

nick-funk opened the pull request:
## What does this PR do?

- premoderate alias email option in config
  - add option on tenant
- users who sign up as an alias of existing user will be pre-modded
- users who sign up as an alias, but NO existing users will NOT be pre-modded
- user signs up with an alias of an account that is banned, will be banned

## These changes will impact:

- [X] commenters
- [X] moderators
- [ ] admins
- [ ] developers

## What changes to the GraphQL/Database Schema does this PR introduce?

Adds premoderate email alias options to `Settings` configs.

## Does this PR introduce any new environment variables or feature flags?

No

## If any indexes were added, were they added to `INDEXES.md`?

N/A

## How do I test this PR?

- enable the feature under `Config > Admin > Moderate`
- create a new user with an alias that has never been used before
  - should not be premodded
- create a user, then create another user that is an alias of that user
  - alias user should be premodded
- create an alias user, then create a second alias user in same vein
  - second user should be premodded
- create a user and ban them
  - create a new user that is an alias of the banned user, they should be banned automatically

## Were any tests migrated to React Testing Library?

No

## How do we deploy this PR?

- Merge into `develop`
- Turn on config option for relevant clients, customers

nick-funk committed:
add config option for premoderating email aliases

nick-funk committed:
add comment about ban check inside the spam premod filter functions

nick-funk committed:
premoderate alias emails if the base of the alias is also a user

nick-funk committed:
ban user aliases if other aliases/base-email user's are banned

nick-funk committed:
make mongo optional in pre-mod filter so tests can run

nick-funk committed:
Merge branch 'develop' into feat/CORL-3213-premoderate-email-aliases

kabeaty commented on client/src/core/client/admin/routes/Configure/sections/Moderation/PremoderateEmailAddressConfig.tsx:
```diff
@@ -62,6 +65,17 @@ const PremoderateEmailAddressConfig: FunctionComponent<Props> = ({
+          <Localized id="configure-moderation-premoderateEmailAliases-enabled">
+            <Label component="legend">Pre-moderate email aliases</Label>
+          </Localized>
+        </FormFieldHeader>
```
just checking - do we have any type of copy for a helper text/description of how this works that we want to add here?

nick-funk replied:
they didn't provide any, so I was going to let Andrew sort it out 😄

tessalt replied:
"If a user signs up for a new account with an email address that is an alias (using a + sign) of an existing account, set their status to pre-moderate comments. Email aliases are commonly used by spammers and trolls to evade bans."

nick-funk replied:
thank you @tessalt!

nick-funk replied:
this has now been updated!

kabeaty replied:
Just checking -- the description helper text that I see here is different from what is above in the thread. Is this correct?

nick-funk replied:
I took what Tessa posted in Slack, I will update to this and hope it's the `text_revision_final_r4` 🤣

nick-funk replied:
it is now updated

kabeaty replied:
sounds good, ty!

kabeaty commented on server/src/core/server/services/users/emailPremodFilter.ts:
```diff
@@ -26,6 +36,33 @@ const emailHasTooManyPeriods = (email: string | undefined, limit: number) => {
+  // mods and admins regularly use aliases for all
+  // their users, so it is less likely they will use
+  // their base email in our system.
+  const existingUser = await mongo
```
hmm, we could also check here the role of the existing user, and only pre-mod if the user is an alias of an existing user with role that is not staff/moderator/admin?

kabeaty commented on server/src/core/server/services/users/emailPremodFilter.ts:
```diff
@@ -71,6 +104,9 @@ export const shouldPremodDueToLikelySpamEmail = (
     return false;
   }
 
+  // if a user is on auto ban list, they will become banned via their
```
thanks for adding this comment here!

kabeaty commented on server/src/core/server/services/users/emailPremodFilter.ts:
```diff
@@ -79,7 +115,15 @@ export const shouldPremodDueToLikelySpamEmail = (
+      EMAIL_PREMOD_FILTER_PERIOD_LIMIT
+    ),
+    // premod email aliases if the feature is enabled
+    tenant?.premoderateEmailAddress?.emailAliases?.enabled && mongo
```
What do you think about moving this check for whether the feature is enabled inside of `emailIsAnAliasOfExistingUser`, similar to how `emailHasTooManyPeriods` does the check?

nick-funk replied:
I had it in there before actually! I move it last night, thinking it made it clear that the check will only run when this feature is enabled, but I can put it back in that function again.

nick-funk replied:
moved it back, is fixed

kabeaty replied:
Lol, sounds good...ty!

kabeaty commented on server/src/core/server/services/users/helpers.ts:
```diff
@@ -104,3 +104,65 @@ export function checkForNewUserEmailDomainModeration(
+    };
+  }
+
+  const domain = split[2];
```
It looks like this and the `aliasStart` assignment could be moved down to line 151-ish or so since we don't utilize `domain` or `aliasStart` until after that?

nick-funk replied:
fixed

kabeaty commented on server/src/core/server/services/users/helpers.ts:
```diff
@@ -104,3 +104,65 @@ export function checkForNewUserEmailDomainModeration(
+    };
+  }
+
+  const split = email.split("@");
```
just wondering if the naming could be clearer here too? Maybe clarifying that this is `emailSplit` would help since there's also the `aliasSplit` below.

nick-funk replied:
fixed

kabeaty commented on server/src/core/server/services/users/helpers.ts:
```diff
@@ -104,3 +104,65 @@ export function checkForNewUserEmailDomainModeration(
+
+  return {
+    base: {
+      value: baseEmail,
```
I am trying to think if there's a way to make these namings any clearer. I had to take some time to think through what each was still. Maybe something like the following could help clarify?:
`baseEmail` --> `baseOfAliasWithDomain`
`aliasStart` --> `aliasWithoutDomain`

nick-funk replied:
fixed

kabeaty reviewed (commented):
Nice! I left a few small comments/questions throughout. I did also get an error when trying to enable the new config and clicking `save changes`, so haven't completed manual testing yet. The error is `Cannot create field 'emailAliases' in element {premoderateEmailAddress: null}`

nick-funk committed:
clean up naming of results from `emailIsAlias` check

nick-funk committed:
move tenant premod filter feature check back into operating func

nick-funk committed:
don't pre-mod existing alias users if they're a staff member

nick-funk committed:
Merge branch 'develop' into feat/CORL-3213-premoderate-email-aliases

nick-funk committed:
lint: formatting of regex

nick-funk committed:
fix: make regex format work with `new RegExp(...)`

- refactored from the `/.../` notation and forgot to
remove the outer `/` when putting as raw string into
`RegExp`
- also updated the local auth flow ban tests to also
include the new alias ban flow if enabled

nick-funk committed:
premod on any prior user, alias or not

updated as per Tessa and Andrews conversations in Slack

nick-funk committed:
add helper text for premod email aliases feature

kabeaty commented on server/src/core/server/services/users/emailAliasBanFilter.ts:
```diff
@@ -0,0 +1,36 @@
+import { MongoContext } from "coral-server/data/context";
+import { parseEmailAliasIntoParts, sanitizeStringForRegex } from "./helpers";
+
+export const shouldBanEmailBecauseOtherAliasesAreBanned = async (
```
I think we need to add a check for premoderate email aliases being enabled in here. I tested out creating a new email alias when another was banned, and with the feature disabled, it still banned the new user.

nick-funk replied:
fixed

kabeaty replied:
hmm I am still seeing my new email alias user being banned by the system when the feature is disabled.

kabeaty replied:
I think `tooManyPeriods` is being checked where `emailAliases` should be.

nick-funk replied:
yeah, that's definitely it, fixed!

kabeaty commented on server/src/core/server/services/users/emailAliasBanFilter.ts:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    if (user.status.ban.active) {
+      return true;
```
I noticed that in the user drawer account history tab in the admin, these bans are described as `Account domain banned.` I think that this was simply the only way that accounts were banned by the system up until now. We might want to either update this text (`moderate-user-drawer-account-history-account-domain-banned`) to clarify that it could be because the account was domain banned or banned for being an alias of a banned email. Or else figure out a way to determine which reason it was banned for.

nick-funk replied:
updated

kabeaty replied:
ty!

kabeaty reviewed (commented):
Looking good overall! Thank you for all the updates. I left a few comments throughout. The only blocking one is that I think we need a check for the premoderate email aliases feature being enabled before banning a user because other aliases are banned.

nick-funk committed:
update premoderate email aliases description text

nick-funk committed:
check alias feature enabled in alias banning flow

nick-funk committed:
updated user history ban text on domains/aliases

nick-funk committed:
fix typo on feature check for premoderateEmailAddress

kabeaty reviewed (approved):
Looks great, ty for the updates!
